cpp/include/cudf/detail/normalizing_iterator.cuh

```diff
@@ -0,0 +1,367 @@
+  input_normalator& operator=(input_normalator&&)      = default;
+
+  /**
+   * @brief Indirection operator returns the value at the current iterator position
+   */
+  __device__ inline Integer operator*() const { return operator[](0); }
+
+  /**
+   * @brief Dispatch functor for resolving a Integer value from any integer type
+   */
+  struct normalize_type {
+    template <typename T, std::enable_if_t<cuda::std::is_integral_v<T>>* = nullptr>
+    __device__ Integer operator()(void const* tp)
+    {
+      return static_cast<Integer>(*static_cast<T const*>(tp));
+    }
+    template <typename T, std::enable_if_t<not cuda::std::is_integral_v<T>>* = nullptr>
+    __device__ Integer operator()(void const*)
+    {
+      CUDF_UNREACHABLE("only integral types are supported");
+    }
+  };
+
+  /**
+   * @brief Array subscript operator returns a value at the input
+   * `idx` position as a `Integer` value.
+   */
+  __device__ inline Integer operator[](size_type idx) const
+  {
+    void const* tp = p_ + (idx * this->width_);
+    return type_dispatcher(this->dtype_, normalize_type{}, tp);
+  }
+
+  /**
+   * @brief Create an input index normalizing iterator.
+   *
+   * Use the indexalator_factory to create an iterator instance.
+   *
+   * @param data      Pointer to an integer array in device memory.
+   * @param data_type Type of data in data
+   */
+  input_normalator(void const* data, data_type dtype)
+    : base_normalator<input_normalator<Integer>, Integer>(dtype), p_{static_cast<char const*>(data)}
+  {
+  }
+
+  char const* p_;  /// pointer to the integer data in device memory
+};
+
+/**
+ * @brief The integer normalizing output iterator
+ *
+ * This is an iterator that can be used for index types (integers) without
+ * requiring a type-specific instance. It can be used for any iterator
+ * interface for writing an array of integer values of type
+ * int8, int16, int32, int64, uint8, uint16, uint32, or uint64.
+ * Setting specific elements always accept the `Integer` type values.
+ *
+ * @tparam Integer The type used for all write functions
+ */
+template <typename Integer>
+struct output_normalator : base_normalator<output_normalator<Integer>, Integer> {
+  friend struct base_normalator<output_normalator<Integer>, Integer>;  // for CRTP
+
+  using reference = output_normalator const&;  // required for output iterators
+
+  output_normalator()                                    = default;
+  output_normalator(output_normalator const&)            = default;
+  output_normalator(output_normalator&&)                 = default;
+  output_normalator& operator=(output_normalator const&) = default;
+  output_normalator& operator=(output_normalator&&)      = default;
+
+  /**
+   * @brief Indirection operator returns this iterator instance in order
+   * to capture the `operator=(size_type)` calls.
+   */
+  __device__ inline output_normalator const& operator*() const { return *this; }
```

Comment:
I'm fairly certain I haven't understood this part. By returning `*this`, then for any `output_normalator` `p`,
`p` == `*p` == `**p`?

Reply:
I would have thought we'd have to support semantics like:
```c++
output_normalator p = ...;
*p = int16_t{5};
*(p+1) = int32_t{10};
p[2] = int64{15};
```
I haven't understood how current implementation for `operator*()` and `operator[]()` work.

Reply:
We cannot return a reference to the underlying array element which is what you would normally see.
From what I remember, this is a way to capture that request `*itr = value` and redirect it to the `operator=(Integer)` function below. It's a subtle thing for sure.

Reply:
The output iterator expects a specific type but can store it as the underlying type.
The `operator[]()` works normally in that it handles `p[2] = value`.
I've tried to explain the `operator*()` and `operator=` relationship in a few separate comments now.

Reply:
Oh, maybe I see it now?

`*p = 5;`  is evaluated as:
1. `output_normalator::operator *()`, which returns `p`.
2. `p = 5` calls `output_normalator::operator =(Integer)`.

Do I have this right?

Were I to nitpick, I would have preferred it if `*p` returned an `integer_reference` object, and `integer_reference::operator=()` did the assignment.  As it stands, doesn't the user run the risk of assigning integers to the normalator directly, instead of to `*normalator`?

That said, yours has the benefit of brevity.

Reply:
That reference idea sounds brilliant to me! I think that’s much clearer in intent.

Reply:
I think there is more to it than what you have described.

*p = 5; is evaluated as:

1. `output_normalator::operator*()`, which returns _reference_ to `p`.
2. `&p = 5` calls `output_normalator::operator =(Integer)`.

I'm not seeing what the `integer_reference` object would be defined to be.

Reply:
Yes, let's do this in a later PR. I'd propose the following. If you agree, opening an issue for this would be good.

- `output_normalator::reference` is a member type defined as a class `integer_reference`
- `integer_reference` requirements:
  - holds some state to know what it's referencing (perhaps the output normalator and the index which it references?)
  - defines `integer_reference::operator=(Integer)` which normalizes and writes to the referenced location
- `output_normalator::operator*` is changed to return an `integer_reference`
- `output_normalator::operator=(Integer)` is no longer defined -- because assignment of a value to a (non-dereferenced) iterator is an unclear design that we can avoid by using `integer_reference`

Reply:
> let's do this in a later PR

Yes, sir. That works.